src/core/license.rs

```diff
@@ -63,82 +96,914 @@ pub enum StateTransition {
+            ))
+        })?;
+
+        let rsa_public_key = match public_key {
+            x509_parser::public_key::PublicKey::RSA(key) => RsaPublicKey::new(
+                rsa::BigUint::from_bytes_be(key.modulus),
+                rsa::BigUint::from_bytes_be(key.exponent),
+            )
+            .map_err(|_| {
+                Error::RdpError(RdpError::new(
+                    RdpErrorKind::InvalidData,
+                    "can't get RSA certificate from the X509 certificate",
+                ))
+            })?,
+            _ => {
+                return Err(Error::RdpError(RdpError::new(
+                    RdpErrorKind::InvalidData,
+                    "invalid type of certificate",
                 )))
             }
+        };
+
+        Ok(Self::X509(rsa_public_key))
+    }
+}
+
+pub struct NewLicense {
+    mac_data: Vec<u8>,
+    encrypted_license_data: Vec<u8>,
+}
+
+impl NewLicense {
+    fn from_bytes(raw: &mut dyn Read) -> RdpResult<Self> {
+        let mut message = component![
+            "EncryptedLicenseInfo" => component![
+                "wBlobType" => U16::LE(0),
+                "wBlobLen" => DynOption::new(U16::LE(0), | size | MessageOption::Size("blobData".to_string(), size.inner() as usize)),
+                "blobData" => Vec::<u8>::new()
+            ],
+            "MACData" => vec![0_u8; 16]
+        ];
+
+        message.read(raw)?;
+        let encrypted_license_data =
+            cast!(DataType::Slice, message["EncryptedLicenseInfo"])?.to_vec();
+        let mac_data = cast!(DataType::Slice, message["MACData"])?.to_vec();
+        Ok(Self {
+            encrypted_license_data,
+            mac_data,
+        })
+    }
+}
+
+pub struct ErrorAlert {
+    code: ErrorCode,
+    state_transition: StateTransition,
+}
+
+impl ErrorAlert {
+    fn from_bytes(raw: &mut dyn Read) -> RdpResult<Self> {
+        let mut message = component![
+        "dwErrorCode" => U32::LE(0),
+        "dwStateTransition" => U32::LE(0),
+        "blob" => component![
+            "wBlobType" => U16::LE(0),
+            "wBlobLen" => DynOption::new(U16::LE(0), | size | MessageOption::Size("blobData".to_string(), size.inner() as usize)),
+            "blobData" => Vec::<u8>::new()
+        ]];
+        message.read(raw)?;
+
+        let code = ErrorCode::try_from(cast!(DataType::U32, message["dwErrorCode"])?)?;
+        let state_transition =
+            StateTransition::try_from(cast!(DataType::U32, message["dwStateTransition"])?)?;
+        Ok(Self {
+            code,
+            state_transition,
+        })
+    }
+
+    fn is_valid(&self) -> RdpResult<()> {
+        if self.code == ErrorCode::StatusValidClient
+            && self.state_transition == StateTransition::StNoTransition
+        {
+            Ok(())
+        } else {
+            Err(Error::RdpError(RdpError::new(
+                RdpErrorKind::InvalidRespond,
+                "server reject license, actually license nego is not implemented",
```

Comment:
a nitpick, but I'm not sure I understand this error message. Is it implying we're missing some functionality on our end, or that the remote RDP server is missing license negotiation?

Reply:
Thanks for catching! So this is the message that has been before I've implemented the license negotiation and I didn't change it to current version. Before it made sense, as that library wasn't handling license negotiation, but now it does not make sense. I've changed it to just say `server reject license`.